adr/ADR-50.md

```diff
@@ -91,7 +90,82 @@ The `LastMsgId` header is currently not supported. A batch will be rejected if t
 
 Initial release of this feature rejects the use of `MsgId`. Starting from 2.12.1 de-duplication is supported and a batch will be rejected with an error if it contains a duplicate message.
 
-### Publish Acknowledgements
+## Fast-ingest Batch Publishing
+
+### Context
+
+Today we have Async publishing in the clients that aims to move data into a stream at high speed, it works well but pays a very high cost for managing acknowledgements both in client and server.
+
+We would like to build on the batching behaviours introduced here but deliver high performance ingest without atomicity.
+
+Crucially these batches are not pre-staged and so will not be limited to 1000 messages like Atomic ones.
+
+### Control Channel
+
+The heart of this design is control channel that is open and kept open for the duration of the batch. In practise this the reply subject of the first message will be used for all signaling during the batch.
+
+The server will send acks over the channel on a frequency like once every 10 messages or once every 5MB. Crucially if at any stage an error is encountered errors can be sent back immediately and received by the client.
+
+Clients will not wait for each ack like they would in a standard JS Publish instead they will maintain a count of maximum outstanding acknowledgements from the server, this is part flow-control and part outage detection.
+
+The server on the other hand will be able to adjust the frequency of acks based on internal metrics of the stream such as the size of the in-flight RAFT proposals. In this way the server and clients will balance speed of fast ingest vs responsiveness of interactive commands requiring RAFT proposals.
+
+At all times the server will maintain its self-protection mechanisms like dropping messages when buffers are full etc.
+
+### Client Design
+
+The client will signal batch start, membership and flow control characteristics using headers on published messages.
+
+ * The client will set up a Inbox subscription that will be used for the duration of the batch
+ * A batch will be started by adding the `Nats-Fast-Batch-Id:uuid`, `Nats-Flow:10` and `Nats-Batch-Sequence:1` headers using a *request* with reply being the above Inbox, the server will reply with error or zero byte message. Maximum length of the ID is 64 characters.
+ * Following messages in the same batch will include the `Nats-Fast-Batch-Id:uuid` header and increment `Nats-Batch-Sequence:n` by one, and might optionally include a reply set to the same Inbox to force a immediate zero byte reply
+ * If the final message has the headers `Nats-Batch-Id:uuid`, `Nats-Batch-Sequence:n` and `Nats-Batch-Commit:1`, the server will store the message, commit the batch and reply with a pub ack.
```

Comment:
The client can continue witha new batch on a new reply subject without waiting for the commit, right?

**If yes: I wonder whats the point of having batching API?** We could tecnically, on the public API side of the server, have a flow-controlled publishing API and server could manage commiting those batches on its own.

I think it does not matter much if that is a gap in single batch or at commit level.

I'm not saying we should do it that way. Just that when reading the ADR that's my immediate thought - that the commiting feels like artifact from `atomic batch` feature.

Reply:
I'd agree the `Nats-Batch-Commit: 1` can be confused with atomic batching. Maybe we should only use `Nats-Batch-Commit: eob` for fast batching, which then doesn't get stored and ends the batch?